### PR TITLE
Properly handle feature-gated lints

### LINKED ISSUE #72694: Properly handle feature-gated lints
At the moment, the `unsafe_op_in_unsafe_fn` lint is the only feature-gated lint. However, feature-gate checks are hard-coded for this lint. If more feature gated lints are be added in the future, having proper handling for feature-gated lint will be useful.

Here is a suggested implementation plan:
- Add a `feature_gate` field of type `Option<Symbol>` for the [`rustc_session::lint::Lint`](https://github.com/rust-lang/rust/blob/45127211566c53bac386b66909a830649182ab7a/src/librustc_session/lint.rs#L57-L88) struct, which has a `Some` value if the lint is feature-gated, `None` otherwise.
- Tweak the [`rustc_session::declare_lint!`](https://github.com/rust-lang/rust/blob/45127211566c53bac386b66909a830649182ab7a/src/librustc_session/lint.rs#L272) macro to allow specifying the feature gate when applicable, akin to `future_incompatible`.
- When registering lints on the command line or for attributes, check if they are feature gated, and report errors accordingly. Basically, re-writing [`check_gated_link`](https://github.com/rust-lang/rust/blob/0e3b31c641217542f02b40527b0d299d9534b920/src/librustc_lint/levels.rs#L389) should be enough.
- As rustdoc allows all lints as if they were passed on the command line, prevent rustdoc from allowing feature gated lints.

This issue has been assigned to @OddCoincidence via [this comment](https://github.com/rust-lang/rust/issues/72694#issuecomment-638628391).

### PR DESCRIPTION
**Pull request by OddCoincidence**
Closes #72694

### CODE REVIEW

File: src/librustc_lint/context.rs [hunk 1]
```diff
@@ -62,6 +63,9 @@ pub struct LintStore {
     /// Lints indexed by name.
     by_name: FxHashMap<String, TargetLint>,
 
+    /// Feature-gated lints.
+    gated_lints: FxHashMap<LintId, (String, Symbol)>,
```
Comment (LeSeulArtichaut) [inline src/librustc_lint/context.rs:67]: This doesn’t seem necessary to me? I think if you have a `LintId`, you can access the associated lint, and then check if the feature gate is enabled.

File: src/librustc_lint/context.rs [hunk 2]
```diff
@@ -275,6 +284,23 @@ impl LintStore {
         }
     }
 
+    /// Check if the lint is gated on a feature that is not enabled.
+    pub fn check_gated_lint(&self, sess: &Session, lint_id: LintId, span: Span) {
+        if let Some((lint_name, feat)) = self
+            .gated_lints
+            .get(&lint_id)
+            .filter(|(_, feat)| !sess.features_untracked().enabled(*feat))
+        {
```
Comment (LeSeulArtichaut) [inline src/librustc_lint/context.rs:293]: Here, I think you would write something like:

```suggestion
        if let Some(feature) = lint_id.lint.feature_gate {
            if !sess.features_untracked().enabled(*feature) {
```

File: src/librustc_lint/context.rs [hunk 3]
```diff
@@ -275,6 +284,23 @@ impl LintStore {
...
+            feature_err(
+                &sess.parse_sess,
+                *feat,
+                span,
+                &format!("the `{}` lint is unstable", lint_name),
```
Comment (LeSeulArtichaut) [inline src/librustc_lint/context.rs:298]: Here you can write:
```suggestion
                &format!("the `{}` lint is unstable", lint_id.lint.name),
```

**Review by LeSeulArtichaut — COMMENTED** [review]
Welcome to the Rust contributors 😄

File: src/librustc_lint/context.rs (hunk 1 above)
Comment (OddCoincidence) [reply src/librustc_lint/context.rs]: Thanks, I missed that the `Lint` could be accessed that way. Fixed this in the latest commit and also moved `check_gated_lint` back to `levels.rs` where it was before.

**Review by LeSeulArtichaut — APPROVED** [review]
Looks good to me now.

**Comment by LeSeulArtichaut** [comment]
The only thing I'm not sure about is whether we should test this behavior, and if we should, how to do it. The only test that exists for now is the feature gate test for `unsafe_block_in_unsafe_fn`, which will eventually be stabilized. Maybe cc @nikomatsakis?

**Comment by petrochenkov** [comment]
Looks like a purely internal refactoring, no special testing should be necessary.

r=me after squashing the commits

**Comment by nikomatsakis** [comment]
Looks good to me too. I don't think further tests are needed. One thing is that it'd be nice to document this but I guess we don't have coverage of creating lints in the rustc-dev-guide that I can see so I guess it's fine for now.

**Comment by OddCoincidence** [comment]
@petrochenkov I squashed the commits as requested

> One thing is that it'd be nice to document this but I guess we don't have coverage of creating lints in the rustc-dev-guide that I can see so I guess it's fine for now.

Perhaps my next contribution could be adding a new lint, and in the process I could add that coverage to the dev guide.

**Comment by petrochenkov** [comment]
Thanks!
@bors r+

**Pull request merged**